### PR TITLE
[Chat] DMs Improvements & Chat Error Handling

### PR DESCRIPTION
**Pull request by rkdavidson**
# Related Todos
- [**🏕 🔥 [DMs] Show indicator for unread/new messages**](https://3.basecamp.com/3926363/buckets/17566548/todos/3005468554)
- [**🏕 🔥 [DMs] New DMs don't show in the list until refresh**](https://3.basecamp.com/3926363/buckets/17566548/todos/3076503485)
- [**🏕 🔥 [Auth] Logging in sometimes throws error**](https://3.basecamp.com/3926363/buckets/17566548/todos/3057764922)
- [**🏕 [Chat] Messages/input visual polish**](https://3.basecamp.com/3926363/buckets/17566548/todos/3076475123) (mostly, at least)
- [**🏕 [DMs] Update DMs banner design**](https://3.basecamp.com/3926363/buckets/17566548/todos/3071488250)

# Changes or Updates
- Major refactor of DMs dropdown to use Bootstrap styled dropdown and show unread indicators
- Listen to message received events to dynamically show new DM's and/or unread indicator
- Hide DM chats where last message received > 12hrs
- Better chat connection error handling, in lieu of more specific solution for intermittent logout/login as different user issue
- MessageInput styling refined
- Disable `/` and `:` and `@` commands for the time being

# Tradeoffs/Caveats

**Responding to new DMs**
This PR uses a heavy handed approach. When a new message is received in a DM channel, we re-fetch all of a users' DM channels. This can be refined later, but for now, is the simplest way to enforce consistency in the various components.

**Chat Error Handling**
Another heavy handed approach, the chat initialization logic is wrapped in a big `try/catch`. I've failed to isolate the _true_ cause of the intermittent Stream Chat errors, but they do seem to only happen when switching from one user to another.
Regardless, to prevent a hard crash of the entire page, a user-friendly error is shown and refreshing the page always fixes it.
This can also be refined later.

# Test Instructions
- Play with direct messages between two users (use an incognito window for second user)
- DMs where the last message received was > 12hrs ago should not appear.
  - For testing, you can change the time value in `src/stream-chat-client/chatUtils.js` on **line 92**
- If someone else initiates a DM with you, it should appear automatically/show the dropdown if it wasn't before
- Unread indicator should show on the **Direct Messages** dropdown as a whole, and on the individual DM channels
- Log out of one user, and in as another without refreshing the page. You _may_ get an error. If you do, verify the site is still usable, and that chat displays a user-friendly message.
- Verify the critical "happy path" for regular users — going from logged out => in state

# Screenshots

|Caption|Screenshot/GIF|
|---|---|
|**DMs Flow**||
|**DMs Dropdown Styling**||
|**Updated MessageInput design**||
|**Chat connection error state**||

# Tests
- [ ] needed snapshots updated
- [ ] mobile tested
- [ ] latest update from master

### CODE REVIEW

**Review by hellogerard — APPROVED** [review]
LGTM! 👍 


Regarding your tradeoffs...

1. I am doing the same thing as you, refreshing DM channels on new messages. Is there any other way to do it?  I remember looking for a "new channel" event or something, but didn't see it. So on a new message, I think the only way to know about a new channel is to fetch them all 🤷 

1. Error handling. I was thinking that Stream.io handled all errors already, which they do, except for the first connection. So I totally stole your error handling idea and put it in my PR.

Lastly, regarding the weird login error, before I was getting it whenever I logged in as one of the Star Wars users. But now I can log in as Anakin and it's working fine. FWIW.

**Comment by rkdavidson** [comment]
> I am doing the same thing as you, refreshing DM channels on new messages. Is there any other way to do it?

This is something I want to look closer at. Since they _heavily_ use custom hooks for the core logic and state management, I'm hoping there's some magic simple way to import one like this [useNotificationAddedToChannelListener](https://github.com/GetStream/stream-chat-react/blob/master/src/components/ChannelList/ChannelList.js#L18) to do a more surgical approach.

Similar problem with updating **unread counts**. Since it's buried in a `channel` instance, which is immutable (or possibly mutable, but shouldn't be), I feel like we'd need to lift that data up and out to track separately (and redundantly).

**TL;DR**, it's a potential can of worms, right?
Definitely solvable, just needs some thinking around.

I'll merge this PR in the morning — don't want to auto deploy when nobody is online.
Thanks for the prompt review @hellogerard 🙇‍♂️.

**Comment by hellogerard** [comment]
> hoping there's some magic simple way to import one like this useNotificationAddedToChannelListener to do a more surgical approach.



Oh, man. I totally never saw the `notification.added_to_channel` event. That's what I get. That would do it. So you'd listen for two events, 1) `notification.added_to_channel` and 2) `notification.message_new`, and handle them separately. You shouldn't need their hooks. You could setup your own listener on `chatClient` or `channel`. 


> Similar problem with updating unread counts. Since it's buried in a channel instance, which is immutable (or possibly mutable, but shouldn't be), I feel like we'd need to lift that data up and out to track separately (and redundantly).


Not sure I follow. If you get a `notification.message_new` event or some other event, then the unread count on the channel should have been just updated.

...To that note, and I'm not totally certain of this, it's just my intuition, I think their JS library holds all the data internally before triggering events or otherwise giving access to app code. That is, if we call a `queryChannels` query, I don't believe it's going to the network or anything. I think it's just executing some local Javascript and returning data already in the app. So there's no real cost. That data will be there whether we listen for it or not.

**Comment by rkdavidson** [comment]
> Not sure I follow. If you get a notification.message_new event or some other event, then the unread count on the channel should have been just updated.

I suppose this is the crux of my problem, @hellogerard — I assumed the same thing, that these magic `channel` objects that get passed around/provided via `context` would be "hot wired" to data changes.
It certainly seems like that's the case, otherwise how would any of this work?

But, perhaps this is _my_ issue with the structure of my custom built React components for web — I'm not seeing the unread indicators update in my `<DirectMessagesDropdown>` list component.

Though, now that I say that... I possibly could (or need to?) use a `React.memo` function to re-render when the `channels` array changes _or_ any individual `unread` values within that array change... 🤔.

Regardless, I 100% agree with your whole comment and am anxious to study their source code a bit closer.
I feel like their usage of hooks is just beyond my current understanding/skills, so I have things to learn.
